credentials/oauth/oauth.go

```diff
@@ -0,0 +1,116 @@
+/*
+ *
+ * Copyright 2014, Google Inc.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+// Package oauth implements grpc credentials using oauth.
+package oauth
+
+import (
+	"fmt"
+	"io/ioutil"
+
+	"google.golang.org/grpc/credentials"
+
+	"golang.org/x/net/context"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
+	"golang.org/x/oauth2/jwt"
+)
+
+// TokenSource supplies credentials from an oauth2.TokenSource.
+type TokenSource struct {
+	oauth2.TokenSource
+}
+
+// GetRequestMetadata gets the request metadata as a map from a TokenSource.
+func (ts TokenSource) GetRequestMetadata(ctx context.Context) (map[string]string, error) {
+	token, err := ts.Token()
+	if err != nil {
+		return nil, err
+	}
+	return map[string]string{
+		"authorization": token.TokenType + " " + token.AccessToken,
+	}, nil
+}
+
+// NewComputeEngine constructs the credentials that fetches access tokens from
+// Google Compute Engine (GCE)'s metadata server. It is only valid to use this
+// if your program is running on a GCE instance.
+// TODO(dsymonds): Deprecate and remove this.
```

Comment:
On Wed, Jul 22, 2015 at 6:41 PM, David Symonds <notifications@github.com>
wrote:

> In credentials/oauth/oauth.go
> <https://github.com/grpc/grpc-go/pull/237#discussion_r35283573>:
>
> > +
> > +// GetRequestMetadata gets the request metadata as a map from a TokenSource.
> > +func (ts TokenSource) GetRequestMetadata(ctx context.Context) (map[string]string, error) {
> > +	token, err := ts.Token()
> > +	if err != nil {
> > +		return nil, err
> > +	}
> > +	return map[string]string{
> > +		"authorization": token.TokenType + " " + token.AccessToken,
> > +	}, nil
> > +}
> > +
> > +// NewComputeEngine constructs the credentials that fetches access tokens from
> > +// Google Compute Engine (GCE)'s metadata server. It is only valid to use this
> > +// if your program is running on a GCE instance.
> > +// TODO(dsymonds): Deprecate and remove this.
>
> @iamqizhao <https://github.com/iamqizhao>: You're missing my point. There
> is a need for credentials.TokenSource where it will *not* be used with
> google.ComputeTokenSource, so we can't unexport it. We can keep
> NewComputeEngine around if you really want; I'm just saying we can't
> unexport TokenSource.
>

What is the problem with
https://github.com/iamqizhao/grpc-go/commit/3104ff998c521fb873ff5e0b8eb12c2bf280f0f8
?

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/grpc/grpc-go/pull/237/files#r35283573>.
>